crates/spirv-std/src/image.rs

```diff
@@ -6,6 +6,62 @@
 mod params;
 
 pub use self::params::{ImageCoordinate, ImageCoordinateSubpassData, SampleType};
+/// A macro for creating SPIR-V `OpTypeImage` types. Always produces a
+/// [`spirv_std::image::Image<...>`](struct@Image) type.
+///
+/// The grammar for the macro is as follows:
+///
+/// ```rust
+/// # // This doctest can't use the normal macro, use an alternative version
+/// # macro_rules! Image {
+/// #   ($($tt: tt)*)=> {}
+/// # }
+/// Image!(
+///     <dimensionality>,
+///     <type=...|format=...>,
+///     [sampled[=<true|false>],]
+///     [multisampled[=<true|false>],]
+///     [arrayed[=<true|false>],]
+///     [depth[=<true|false>],]
+/// )
+/// # ;
+/// ```
+///
+/// `=true` can be omitted as shorthand - e.g. `sampled` is short for `sampled=true`.
+///
+/// A basic example looks like this:
+/// ```rust,no_run
+/// # // This doctest should compile, but we disable doctests for this crate anyway
+/// # // because this single test isn't worth the compile time cost
+/// # use spirv_std::macros::{spirv, Image};
+/// #[spirv(vertex)]
+/// fn main_vs(#[spirv(descriptor_set = 0, binding = 0)] image: &Image!(2D, type=f32, sampled)) {}
+/// ```
+///
+/// ## Arguments
+///
+/// - `dimensionality` — Dimensionality of an image.
+///    Accepted values: `1D`, `2D`, `3D`, `rect`, `cube`, `subpass`.
+/// - `type` — The sampled type of an image, mutually exclusive with `format`,
+///    when set the image format is unknown.
+///    Accepted values: `f32`, `f64`, `u8`, `u16`, `u32`, `u64`, `i8`, `i16`, `i32`, `i64`.
+/// - `format` — The image format of the image, mutually exclusive with `type`.
+///    Accepted values: Snake case versions of [`ImageFormat`].
+/// - `sampled` — Whether it is known that the image will be used with a sampler.
+///    Accepted values: `true` or `false`. Default: `unknown`.
+/// - `multisampled` — Whether the image contains multisampled content.
+///    Accepted values: `true` or `false`. Default: `false`.
+/// - `arrayed` — Whether the image contains arrayed content.
+///    Accepted values: `true` or `false`. Default: `false`.
+/// - `depth` — Whether it is known that the image is a depth image.
+///    Accepted values: `true` or `false`. Default: `unknown`.
+///
+/// [`ImageFormat`]: spirv_types::image_params::ImageFormat
+///
+/// Keep in mind that `sampled` here is a different concept than the `SampledImage` type:
+/// `sampled=true` means that this image requires a sampler to be able to access, while the
+/// `SampledImage` type bundles that sampler together with the image into a single type (e.g.
+/// `sampler2D` in GLSL, vs. `texture2D`).
```

Comment:
The text for the original macro definition is bizarrely appended to this text. That should be figured out why before merging.

Reply:
Ah, interesting

I only added that doc comment as a last minute thing before I had to go again

I think the best way is just to turn it into a normal comment.

Reply:
A normal comment would be problematic, as it wouldn't show up in the macro crate's docs.

Reply:
How much do we actually care about the macro crate docs? 

That is, would we be better off serving all of the macros as `spirv_std::whatever`, e.g. by using `#[doc(inline)]` on re-exports?